api/sip.py

```diff
@@ -0,0 +1,596 @@
+        # fixed-width fields, and build a dictionary of named fields
+        # to use later.
+        for field in fields:
+            if isinstance(field, fixed):
+                data = field.consume(data, parsed)
+            else:
+                fields_by_sip_code[field.sip_code] = field
+                if field.req:
+                    required_fields_not_seen.add(field)
+                
+        # We now have a list of named fields separated by
+        # self.separator.  Use separator_re to split the data in a way
+        # that minimizes the chances that embedded separators (which
+        # shouldn't happen, but do) don't ruin the data.
+        split = self.separator_re.split(data)
+
+        # We now have alternating field name/value pairs, except for
+        # the first field, which wasn't split because it didn't start with
+        # the separator. Fix that.
+        first_field = split[0]
+        first_field = [first_field[:2], first_field[2:]]
+        split = first_field + split[1:]
+        i = 0
+
+        # Now go through each name/value pair, find the corresponding
+        # field object, and process it.
+        while i < len(split):
+            sip_code = split[i]
+            value = split[i+1]
+            if sip_code == named.sequence_number.sip_code:
+                # Sequence number is special in two ways. First, it
+                # indicates the end of the message. Second, it doesn't
+                # have to be explicitly mentioned in the list of
+                # fields -- we always expect it.
+                named.sequence_number.consume(value, parsed)
+                break
+            else:
+                field = fields_by_sip_code.get(sip_code)
+                if not field:
+                    # This is an extension field. Do the best we can.
+                    # This basically means storing it in the dictionary
+                    # under its SIP code.
+                    field = named(sip_code, sip_code, allow_multiple=True)
+
+                field.consume(value, parsed)
+                if field.required and field in required_fields_not_seen:
+                    required_fields_not_seen.remove(field)
+            i += 2
+            field = fields_by_sip_code
+                
+        # If a named field is required and never showed up, sound the alarm.
+        for field in required_fields_not_seen:
+            self.log.error(
+                "Expected required field %s but did not find it.",
+                field.sip_code
+            )
+        return parsed
+    
+    def consume_status_code(self, data, expected, in_progress):
+        """Pull the status code (the first two characters) off the 
+        given response string, and verify that it's as expected.
+        """
+        status_code = data[:2]
+        in_progress['_status'] = status_code
+        if status_code != expected:
+            raise ValueError(
```

Comment:
Should it always be a value error?  Would we want to have different handling for different http statuses, or for when the SC got logged out, or when ACS asks for a message resend or responds to a previous message, rather than the current one?